### PR TITLE
Rework Mapper as type class, introduce ToTwitterFuture type class

### PR DESCRIPTION
**Pull request by ImLiar**
Mapper now is not a magnet pattern, but a type class. Signature of `EndpointMapper.apply` has changed a little, but it's backward compatible. Strictly speaking, from API point of view nothing was changed, except that IDE should stop whining about "wrong" arguments: https://github.com/finagle/finch/issues/631

At the same time, added `ToTwitterFuture` type class to support different async mechanisms. There is an identity for twitter future and scala futures support in different package. Inspired by https://github.com/finagle/finch/pull/815

Potentially, this PR could open a way to support monix and scalaz tasks as parts of different subprojects: https://github.com/finagle/finch/pull/815#issuecomment-318662893

It's a working code, but without specs. If it's valid from "ideological" PoV I could write proper tests for Mapper.

### CODE REVIEW

**Comment by vkostyukov** [comment]
@ImLiar Thanks for working on this! Have you checked IntelliJ to see if it complains about endpoints now?

**Comment by ImLiar** [comment]
screen shot 2017-08-08 at 00 58 08

**Comment by vkostyukov** [comment]
That's really nice! Let me pull this locally and play with it a bit.

File: core/src/main/scala/io/finch/syntax/EndpointMapper.scala [hunk 1]
```diff
@@ -8,7 +8,7 @@ class EndpointMapper[A](m: Method, e: Endpoint[A]) extends Endpoint[A] { self =>
   /**
    * Maps this endpoint to either `A => Output[B]` or `A => Future[Output[B]]`.
    */
-  final def apply(mapper: Mapper[A]): Endpoint[mapper.Out] = mapper(self)
+  final def apply[F](f: => F)(implicit mapper: Mapper[F, A]): Endpoint[mapper.Out] = mapper(f, self)
```
Comment (vkostyukov) [inline core/src/main/scala/io/finch/syntax/EndpointMapper.scala]: The only concerning bit is this by-name value. I understand why we need it, but I wish we could do better than it. I'm worried about an overhead (for a wrapping function0) for a general case (map `A => B`).

Comment (ImLiar) [inline core/src/main/scala/io/finch/syntax/EndpointMapper.scala]: Spent a day, trying to find possible workaround. I wish we would have Dotty now -__-
Seems like magnet is the only possible choice here if there is a need of that kind of optimization. I could move `ToTwitterFuture` stuff to different PR

Comment (vkostyukov) [reply core/src/main/scala/io/finch/syntax/EndpointMapper.scala]: I think we can keep `ToTwitterFuture`. I actually like it.

File: core/src/main/scala/io/finch/syntax/Mapper.scala [hunk 2]
```diff
@@ -13,103 +13,114 @@ import shapeless.ops.function.FnToProduct
...
 private[finch] trait LowPriorityMapperConversions {
-  type Aux[A, B] = Mapper[A] { type Out = B }
+  type Aux[F, A, B] = Mapper[F, A] { type Out = B }
 
-  def instance[A, B](f: Endpoint[A] => Endpoint[B]): Mapper.Aux[A, B] = new Mapper[A] {
+  def instance[F, A, B](fn: (Endpoint[A], => F) => Endpoint[B]): Mapper.Aux[F, A, B] = new Mapper[F, A] {
     type Out = B
-    def apply(e: Endpoint[A]): Endpoint[B] = f(e)
+
+    override def apply(f: => F, e: Endpoint[A]): Endpoint[B] = fn(e, f)
```
Comment (vkostyukov) [inline core/src/main/scala/io/finch/syntax/Mapper.scala]: You don't need an `override` here.

File: core/src/main/scala/io/finch/syntax/Mapper.scala [hunk 3]
```diff
@@ -13,103 +13,114 @@ import shapeless.ops.function.FnToProduct
...
   }
 
   /**
     * @group LowPriorityMapper
     */
-  implicit def mapperFromOutputFunction[A, B](f: A => Output[B]): Mapper.Aux[A, B] =
-    instance(_.mapOutput(f))
+  implicit def mapperFromOutputFunction[A, B]: Mapper.Aux[A => Output[B], A, B] =
+    instance(_ mapOutput _)
```
Comment (vkostyukov) [inline core/src/main/scala/io/finch/syntax/Mapper.scala]: Lets use named args: `instance((e, f) => e.mapOutput(f))`.

File: core/src/main/scala/io/finch/syntax/Mapper.scala [hunk 4]
```diff
@@ -13,103 +13,114 @@ import shapeless.ops.function.FnToProduct
...
 
   /**
    * @group LowPriorityMapper
    */
-  implicit def mapperFromResponseFunction[A](f: A => Response): Mapper.Aux[A, Response] =
-    instance(_.mapOutput(f.andThen(r => Output.payload(r, r.status))))
+  implicit def mapperFromResponseFunction[A]: Mapper.Aux[A => Response, A, Response] =
+    instance(_ mapOutput _.andThen(r => Output.payload(r, r.status)))
```
Comment (vkostyukov) [inline core/src/main/scala/io/finch/syntax/Mapper.scala]: Use named arguments, please.

File: core/src/main/scala/io/finch/syntax/Mapper.scala [hunk 5]
```diff
@@ -13,103 +13,114 @@ import shapeless.ops.function.FnToProduct
...
 
   /**
     * @group LowPriorityMapper
     */
-  implicit def mapperFromFutureOutputFunction[A, B](f: A => Future[Output[B]]): Mapper.Aux[A, B] =
-    instance(_.mapOutputAsync(f))
+  implicit def mapperFromHOutputFunction[F[_], A, B](implicit
+                                                     ttf: ToTwitterFuture[F]
+                                                    ): Mapper.Aux[A => F[Output[B]], A, B] =
+    instance(_ mapOutputAsync _.andThen(ttf.apply))
```
Comment (vkostyukov) [inline core/src/main/scala/io/finch/syntax/Mapper.scala]: I'd prefer to not rely on `_` so much - it's hard to parse.

File: core/src/main/scala/io/finch/syntax/Mapper.scala [hunk 6]
```diff
@@ -13,103 +13,114 @@ import shapeless.ops.function.FnToProduct
...
 
   /**
    * @group LowPriorityMapper
    */
-  implicit def mapperFromFutureResponseFunction[A](f: A => Future[Response]): Mapper.Aux[A, Response] =
-    instance(_.mapOutputAsync(f.andThen(fr => fr.map(r => Output.payload(r, r.status)))))
+  implicit def mapperFromHResponseFunction[F[_], A](implicit
+    ttf: ToTwitterFuture[F]
+  ): Mapper.Aux[A => F[Response], A, Response] =
+    instance(_ mapOutputAsync _.andThen(ttf.apply).andThen(fr => fr.map(r => Output.payload(r, r.status))))
```
Comment (vkostyukov) [inline core/src/main/scala/io/finch/syntax/Mapper.scala]: Again: use named args.

File: core/src/main/scala/io/finch/syntax/Mapper.scala [hunk 7]
```diff
@@ -13,103 +13,114 @@ import shapeless.ops.function.FnToProduct
...
-    instance(_.mapOutput(_ => o))
+  implicit def mapperFromOutputValue[A]: Mapper.Aux[Output[A], HNil, A] =
+    instance((e, o) => e.mapOutput(_ => o))
 
   /**
    * @group HighPriorityMapper
    */
-  implicit def mapperFromResponseValue(r: => Response): Mapper.Aux[HNil, Response] =
-    instance(_.mapOutput(_ => Output.payload(r, r.status)))
+  implicit def mapperFromResponseValue: Mapper.Aux[Response, HNil, Response] =
```
Comment (vkostyukov) [inline core/src/main/scala/io/finch/syntax/Mapper.scala]: Could be `val`.

File: core/src/main/scala/io/finch/syntax/ToTwitterFuture.scala [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+trait ToTwitterFuture[F[_]] {
+
+  def apply[A](f: F[A]): Future[A]
+
+}
+
+object ToTwitterFuture {
+
+  implicit val identity: ToTwitterFuture[Future] = new ToTwitterFuture[Future] {
+    override def apply[A](f: Future[A]): Future[A] = f
```
Comment (vkostyukov) [inline core/src/main/scala/io/finch/syntax/ToTwitterFuture.scala]: No need for `override`.

File: core/src/main/scala/io/finch/syntax/ToTwitterFuture.scala [hunk 9]
```diff
@@ -0,0 +1,17 @@
+package io.finch.syntax
+
+import com.twitter.util.Future
+
+trait ToTwitterFuture[F[_]] {
```
Comment (vkostyukov) [inline core/src/main/scala/io/finch/syntax/ToTwitterFuture.scala]: Please add some minimal scaladoc.

File: core/src/main/scala/io/finch/syntax/EndpointMapper.scala (hunk 1 above)
Comment (rpless) [reply core/src/main/scala/io/finch/syntax/EndpointMapper.scala]: @vkostyukov are you still concerned with the overhead of the by-name parameter? If so maybe a few of us could run the benchmarks against this branch and master to get an idea of how much it adds?

Comment (vkostyukov) [reply core/src/main/scala/io/finch/syntax/EndpointMapper.scala]: Yeah. I think it's a reasonable trade-off to make: better user experience (fewer highlighting erros) > +16 byte increase in allocations per endpoint. But yeah - running some benchmarks would be a good idea.

File: core/src/main/scala/io/finch/syntax/ToTwitterFuture.scala (hunk 8 above)
Comment (ImLiar) [reply core/src/main/scala/io/finch/syntax/ToTwitterFuture.scala]: Thanks for comments!

I was playing around with testing of this beast, but went way too far with scalacheck generators. Going to try again

File: core/src/main/scala/io/finch/syntax/Mapper.scala [hunk 10]
```diff
@@ -1,7 +1,8 @@
 package io.finch.syntax
 
+import _root_.scala.annotation.implicitNotFound
```
Comment (vkostyukov) [inline core/src/main/scala/io/finch/syntax/Mapper.scala]: Do we need `_root_`?

File: core/src/main/scala/io/finch/syntax/EndpointMapper.scala [hunk 11]
```diff
@@ -2,13 +2,14 @@ package io.finch.syntax
 
 import com.twitter.finagle.http.Method
 import io.finch._
+import shapeless.Lazy
```
Comment (vkostyukov) [inline core/src/main/scala/io/finch/syntax/EndpointMapper.scala]: Unused?

File: core/src/main/scala/io/finch/syntax/Mapper.scala [hunk 12]
```diff
@@ -13,103 +14,126 @@ import shapeless.ops.function.FnToProduct
...
+  """"A value you're trying to apply to Endpoint is missing Mapper instance.
+
+  Make sure ${A} is one of the following:
+
+  * A com.twitter.finagle.http.Response or io.finch.Output
+  * A com.twitter.util.Future[Response] (or Output)
+  * A value of type F[Response] (or Output) with implicit `io.finch.syntax.ToTwitterFuture` instance
+  * A function with sufficient amount of arguments that returns one of values above
+
+  See https://github.com/finagle/finch/blob/master/docs/cookbook.md#mapper-syntax
```
Comment (vkostyukov) [inline core/src/main/scala/io/finch/syntax/Mapper.scala]: This doesn't exist yet?

Comment (ImLiar) [reply core/src/main/scala/io/finch/syntax/Mapper.scala]: Yep, just a placeholder for now.

File: core/src/main/scala/io/finch/syntax/Mapper.scala (hunk 10 above)
Comment (ImLiar) [reply core/src/main/scala/io/finch/syntax/Mapper.scala]: There is a `scala` package in a same scope, so yeah

Comment (vkostyukov) [reply core/src/main/scala/io/finch/syntax/Mapper.scala]: Ugh. Right.

File: core/src/main/scala/io/finch/syntax/Mapper.scala [hunk 13]
```diff
@@ -13,103 +14,126 @@ import shapeless.ops.function.FnToProduct
...
+  implicit def mapperFromResponseFunction[A]: Mapper.Aux[A => Response, A, Response] =
+    instance((e, f) => e.mapOutput(f.andThen(r => Output.payload(r, r.status))))
 
   /**
     * @group LowPriorityMapper
     */
-  implicit def mapperFromFutureOutputFunction[A, B](f: A => Future[Output[B]]): Mapper.Aux[A, B] =
-    instance(_.mapOutputAsync(f))
+  implicit def mapperFromHOutputFunction[F[_], A, B](implicit
+                                                     ttf: ToTwitterFuture[F]
```
Comment (vkostyukov) [inline core/src/main/scala/io/finch/syntax/Mapper.scala]: This probably needs formatting to be fixed.

File: core/src/main/scala/io/finch/syntax/EndpointMapper.scala (hunk 1 above)
Comment (vkostyukov) [reply core/src/main/scala/io/finch/syntax/EndpointMapper.scala]: I think we can add a new benchmark under `benchmarks` project.

**Comment by vkostyukov** [comment]
I think this would also resolve #383.

**Comment by ImLiar** [comment]
Let me add tests here first anyway

**Comment by vkostyukov** [comment]
Hey @ImLiar! How is this going? Do you need any help?

**Comment by ImLiar** [comment]
Hi

I'm on shorty vacation for few days. Going to finish on Sunday

**Comment by ImLiar** [comment]
@vkostyukov @rpless 

Benchmark results for this PR and master:

```
PR:
Benchmark                               Mode  Cnt    Score     Error  Units
ApplyBenchmark.function1FutureOutput    avgt   10   61.254 ±  10.093  ns/op
ApplyBenchmark.function1FutureResponse  avgt   10  110.490 ±  73.496  ns/op
ApplyBenchmark.function1Output          avgt   10  116.603 ± 107.588  ns/op
ApplyBenchmark.function1Response        avgt   10   71.227 ±  23.770  ns/op
ApplyBenchmark.functionNFutureOutput    avgt   10  142.614 ±  89.723  ns/op
ApplyBenchmark.functionNFutureResponse  avgt   10  116.035 ±  88.242  ns/op
ApplyBenchmark.functionNOutput          avgt   10   82.280 ±   9.138  ns/op
ApplyBenchmark.functionNResponse        avgt   10  135.343 ±  19.729  ns/op
ApplyBenchmark.futureOutput             avgt   10    7.761 ±   0.613  ns/op
ApplyBenchmark.futureResponse           avgt   10    7.410 ±   0.192  ns/op
ApplyBenchmark.output                   avgt   10    7.609 ±   0.606  ns/op
ApplyBenchmark.response                 avgt   10    7.403 ±   0.229  ns/op

Master:
Benchmark                               Mode  Cnt    Score    Error  Units
ApplyBenchmark.function1FutureOutput    avgt   10   66.755 ± 12.514  ns/op
ApplyBenchmark.function1FutureResponse  avgt   10   68.049 ± 11.998  ns/op
ApplyBenchmark.function1Output          avgt   10   66.912 ± 14.796  ns/op
ApplyBenchmark.function1Response        avgt   10   95.094 ± 56.855  ns/op
ApplyBenchmark.functionNFutureOutput    avgt   10  124.046 ± 33.424  ns/op
ApplyBenchmark.functionNFutureResponse  avgt   10   76.451 ± 13.739  ns/op
ApplyBenchmark.functionNOutput          avgt   10   82.871 ±  8.452  ns/op
ApplyBenchmark.functionNResponse        avgt   10  148.733 ± 30.853  ns/op
ApplyBenchmark.futureOutput             avgt   10    7.430 ±  0.117  ns/op
ApplyBenchmark.futureResponse           avgt   10    7.338 ±  0.192  ns/op
ApplyBenchmark.output                   avgt   10    7.738 ±  0.749  ns/op
ApplyBenchmark.response                 avgt   10    7.849 ±  0.728  ns/op

PR:
Benchmark                                                             Mode  Cnt     Score     Error   Units
ApplyBenchmark.function1FutureOutput:·gc.alloc.rate.norm              avgt   10   188.000 ±  19.124    B/op
ApplyBenchmark.function1FutureResponse:·gc.alloc.rate.norm            avgt   10   188.000 ±  19.124    B/op
ApplyBenchmark.function1Output:·gc.alloc.rate.norm                    avgt   10   200.000 ±   0.001    B/op
ApplyBenchmark.function1Response:·gc.alloc.rate.norm                  avgt   10   176.000 ±   0.001    B/op
ApplyBenchmark.functionNFutureOutput:·gc.alloc.rate.norm              avgt   10   168.000 ±   0.001    B/op
ApplyBenchmark.functionNFutureResponse:·gc.alloc.rate.norm            avgt   10   168.000 ±   0.001    B/op
ApplyBenchmark.functionNOutput:·gc.alloc.rate.norm                    avgt   10   168.000 ±   0.001    B/op
ApplyBenchmark.functionNResponse:·gc.alloc.rate.norm                  avgt   10   420.000 ±  19.124    B/op
ApplyBenchmark.futureOutput:·gc.alloc.rate.norm                       avgt   10    ≈ 10⁻⁵              B/op
ApplyBenchmark.futureResponse:·gc.alloc.rate.norm                     avgt   10    ≈ 10⁻⁵              B/op
ApplyBenchmark.output:·gc.alloc.rate.norm                             avgt   10    ≈ 10⁻⁵              B/op
ApplyBenchmark.response:·gc.alloc.rate.norm                           avgt   10    ≈ 10⁻⁵              B/op

Master:
Benchmark                                                             Mode  Cnt     Score     Error   Units
ApplyBenchmark.function1FutureOutput:·gc.alloc.rate.norm              avgt   10   188.000 ±  19.124    B/op
ApplyBenchmark.function1FutureResponse:·gc.alloc.rate.norm            avgt   10   176.000 ±   0.001    B/op
ApplyBenchmark.function1Output:·gc.alloc.rate.norm                    avgt   10   188.000 ±  19.124    B/op
ApplyBenchmark.function1Response:·gc.alloc.rate.norm                  avgt   10   188.000 ±  19.124    B/op
ApplyBenchmark.functionNFutureOutput:·gc.alloc.rate.norm              avgt   10   168.000 ±   0.001    B/op
ApplyBenchmark.functionNFutureResponse:·gc.alloc.rate.norm            avgt   10   168.000 ±   0.001    B/op
ApplyBenchmark.functionNOutput:·gc.alloc.rate.norm                    avgt   10   168.000 ±   0.001    B/op
ApplyBenchmark.functionNResponse:·gc.alloc.rate.norm                  avgt   10   412.000 ±   6.375    B/op
ApplyBenchmark.futureOutput:·gc.alloc.rate.norm                       avgt   10    ≈ 10⁻⁵              B/op
ApplyBenchmark.futureResponse:·gc.alloc.rate.norm                     avgt   10    ≈ 10⁻⁵              B/op
ApplyBenchmark.output:·gc.alloc.rate.norm                             avgt   10    ≈ 10⁻⁵              B/op
ApplyBenchmark.response:·gc.alloc.rate.norm                           avgt   10    ≈ 10⁻⁵              B/op
```

So there is a quite a difference in overhead. By the way, while working with this PR, I realised that my latest IDEA highlighting works correctly with master for every case but gives an error for `Future[Output[A]]` case in PR... Damn tools

How about postpone decision about this one and move `ToTwitterFuture` to a separate PR? It would open a door for different async libraries meanwhile

**Comment by vkostyukov** [comment]
@ImLiar What's the plan on this? Do we want to continue working in that direction or not?

**Comment by ImLiar** [comment]
@vkostyukov hey

I think I could finish this one, but it won't help 100% with highlighting.
